### PR TITLE
Native Genode target

### PR DESCRIPTION
**Pull request by ghost**
Support for a "genode" target that uses Solo5 to run as a native Genode
component.

---
I am writing an article that will explain how to build a Genode boot image locally, and we are still working out how to do remote deployment on an already running server.

Ref https://github.com/genodelabs/genode/issues/2945

### CODE REVIEW

**Comment by ehmry** [comment]
Well right now it should be possible to build a Mirage image for Genode without anything from us, but to actually use it, you need our scripts and metadata, which are on a topic branch of mine. I would say yes, merge it for the release, and then I can write wrappers over git and mirage so our build system can do the build and integration stuff.

Our next quarterly release is at the end of November (this month) and we would like to mention Mirage in the release.

File: lib/mirage.ml [hunk 1]
```diff
@@ -688,7 +688,7 @@ let extra_c_artifacts target pkgs =
   in
   R.ok r
 
-let static_libs pkg_config_deps = pkg_config pkg_config_deps [ "--static" ; "--libs" ]
+let static_libs pkg_config_deps = pkg_config pkg_config_deps [ "--static" ; "--libs"]
```
Comment (mato) [inline lib/mirage.ml]: Spurious whitespace change.

**Comment by hannesm** [comment]
thanks, sounds good to me. and also your PR looks good to me! :)

**Comment by samoht** [comment]
LGTM

At one point, we should probably refactor the solo5 targets into their own variant (e.g. `| 'Solo5 of ['Virtio|....|'Genode]) at one point, but can be done in another PR.

File: lib/mirage_impl_block.ml [hunk 2]
```diff
@@ -68,7 +68,7 @@ class block_conf file =
     method! packages =
       Key.match_ Key.(value target) @@ function
       | `Xen | `Qubes -> xen_block_packages
-      | `Virtio | `Hvt | `Muen -> [ package ~min:"0.3.0" "mirage-block-solo5" ]
+      | `Virtio | `Hvt | `Muen | `Genode -> [ package ~min:"0.3.0" "mirage-block-solo5" ]
```
Comment (hannesm) [inline lib/mirage_impl_block.ml]: line break after `->` please

**Comment by hannesm** [comment]
(please ignore my nits, i cherry-picked, fixed some more things in mirage, and currently have a CI run..)

File: lib/mirage.ml (hunk 1 above)
Comment (ehmry) [reply lib/mirage.ml]: Genode is linked shared, so I added a `shared_libs` function, but it makes no difference, so I manually reverted and the whitespace change snuck in.

**Comment by ehmry** [comment]
Just to be clear, to make this work I am still pinning everything from solo5-bindings-genode to mirage, and I have to manually edit `mirage-unikernel-*-genode.opam` to remove the version number from the "solo5-bindings-genode" dependency.

```
# opam pin
mirage.3.2.0                                git  git+file:///home/repo/mirage/mirage#genode
mirage-runtime.3.2.0                        git  git+file:///home/repo/mirage/mirage#genode
mirage-solo5.0.4.0                          git  git+file:///home/repo/mirage/mirage-solo5#master
mirage-types.3.2.0                          git  git+file:///home/repo/mirage/mirage#genode
mirage-types-lwt.3.2.0                      git  git+file:///home/repo/mirage/mirage#genode
ocaml-freestanding.0.4.0                    git  git+file:///home/repo/mirage/ocaml-freestanding#master
solo5-bindings-genode.~dev                  git  git+file:///home/repo/mirage/solo5#master
solo5-bindings-hvt.0.4.0     (uninstalled)  git  git+file:///home/repo/mirage/solo5#master
solo5-bindings-muen.0.4.0    (uninstalled)  git  git+file:///home/repo/mirage/solo5#master
solo5-bindings-virtio.0.4.0  (uninstalled)  git  git+file:///home/repo/mirage/solo5#master
```

**Comment by mato** [comment]
Yes, that makes sense. By the way, you can pin with a version number, so, e.g.

    opam pin add solo5-bindings-genode.0.5.0 ...

That way you shouldn't have to change anything.

**Comment by mato** [comment]
Also note that due to mirage/mirage-solo5#37 (in progress) you may need to pin mirage-{block,net}-solo5#master in the interim (i.e. next 24hrs or so)

**Comment by mato** [comment]
Note that at some point we should add `genode` targets to the various CIs (mirage, mirage-skeleton, mirage-solo5, ...). However, in order to keep my sanity in the "maze of versions all alike", I'd leave that until after Mirage 3.3.0 is released.

**Pull request merged**